pkg/build/api/validation/validation.go

```diff
@@ -36,12 +40,38 @@ func ValidateBuildUpdate(build *buildapi.Build, older *buildapi.Build) field.Err
 		allErrs = append(allErrs, field.Invalid(field.NewPath("status", "phase"), build.Status.Phase, "phase cannot be updated from a terminal state"))
 	}
 	if !kapi.Semantic.DeepEqual(build.Spec, older.Spec) {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), "content of spec is not printed out, please refer to the \"details\"", "spec is immutable"))
+		diff, err := diffBuildSpec(build.Spec, older.Spec)
+		if err != nil {
+			glog.V(2).Infof("Error calculating build spec patch: %v", err)
+			diff = "[unknown]"
+		}
+		detail := fmt.Sprintf("spec is immutable, diff: %s", diff)
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), "content of spec is not printed out, please refer to the details", detail))
 	}
 
 	return allErrs
 }
 
+func diffBuildSpec(newer buildapi.BuildSpec, older buildapi.BuildSpec) (string, error) {
+	codec := kapi.Codecs.LegacyCodec(buildapi.SchemeGroupVersion)
+	newerObj := &buildapi.Build{Spec: newer}
+	olderObj := &buildapi.Build{Spec: older}
+
+	newerJSON, err := runtime.Encode(codec, newerObj)
```

Comment:
Specifying the version here ought to encode you to a version.  You'll need need the obj you pass to `CreateTwoWayMergePatch` to match version-wise.